veps/sig-compute/10-dra-devices/vep.md

```diff
@@ -0,0 +1,760 @@
+1. the resourceclaim status has device name and driver use for allocating the device,
+   resourceclaim.status.allocation.devices[*].deviceName and resourceclaim.status.allocation.devices[*].driver, where
+   resourceclaim.status.allocation.devices[*].request is same as vmi.spec.domain.devices[*].gpus[*].claim.request
+1. Using node name and driver name, the resource slice for that node could be found. Using device name, the attributes
+   of the device could be found
+
+```
+---
+apiVersion: v1
+kind: Pod
+metadata:
+  name: virt-launcher-vmi-fedora-9bjwb
+  namespace: gpu-test1
+spec:
+  containers:
+  - name: compute
+    resources:
+      claims:
+      - name: gpu-resource-claim
+  resourceClaims:
+  - name: gpu-resource-claim
+    resourceClaimTemplateName: single-gpu
+status:
+  resourceClaimStatuses:
+  - name: gpu-resource-claim
+    resourceClaimName: virt-launcher-vmi-fedora-9bjwb-gpu-resource-claim-m4k28
+---
+apiVersion: resource.k8s.io/v1alpha3
+kind: ResourceClaim
+metadata:
+  annotations:
+    resource.kubernetes.io/pod-claim-name: gpu-resource-claim
+  generateName: virt-launcher-vmi-fedora-9bjwb-gpu-resource-claim-
+  name: virt-launcher-vmi-fedora-9bjwb-gpu-resource-claim-m4k28
+  namespace: gpu-test1
+  ownerReferences:
+  - apiVersion: v1
+    blockOwnerDeletion: true
+    controller: true
+    kind: Pod
+    name: virt-launcher-vmi-fedora-9bjwb
+spec:
+  devices:
+    requests:
+    - allocationMode: ExactCount
+      count: 1
+      deviceClassName: gpu.example.com
+      name: gpu
+status:
+  allocation:
+    devices:
+      results:
+      - device: pgpu-0
+        driver: gpu.example.com
+        pool: kind-1.31-dra-control-plane
+        request: gpu
+    nodeSelector:
+      nodeSelectorTerms:
+      - matchFields:
+        - key: metadata.name
+          operator: In
+          values:
+          - kind-1.31-dra-control-plane
+  reservedFor:
+  - name: virt-launcher-vmi-fedora-9bjwb
+    resource: pods
+    uid: 8ffb7e04-6c4b-4fc7-bbaa-c60d9a1e0eaa
+---
+apiVersion: resource.k8s.io/v1alpha3
+kind: ResourceSlice
+metadata:
+  generateName: kind-1.31-dra-control-plane-gpu.example.com-
+  name: kind-1.31-dra-control-plane-gpu.example.com-drr27
+  ownerReferences:
+  - apiVersion: v1
+    controller: true
+    kind: Node
+    name: kind-1.31-dra-control-plane
+spec:
+  devices:
+  - basic:
+      attributes:
+        driverVersion:
+          version: 1.0.0
+        index:
+          int: 0
+        model:
+          string: LATEST-GPU-MODEL
+        uuid:
+          string: gpu-8e942949-f10b-d871-09b0-ee0657e28f90
+        pciAddress:
+          string: 0000:01:00.0 
+    name: pgpu-0
+  driver: gpu.example.com
+  nodeName: kind-1.31-dra-control-plane
+  pool:
+    generation: 0
+    name: kind-1.31-dra-control-plane
+    resourceSliceCount: 1
+---
+```
+
+### Web hook changes
+
+1. Allow DRA devices to be requested only if the corresponding DRA feature gate is enabled in kubevirt configuration
```

Comment:
Not sure if it makes sense to check for the FG in a webhook. In my opinion, VMs with DRA stuff defined should instead fail to start if the FG is missing. That would also cover cases where the FG is set/unset after VM creation.

Reply:
It will be really challenging to support dynamic turning on and turning off of feature gates. 

If a user turns on an alpha feature gate, it is reasonable in my opinion to ask them to delete the resources they used with the feature gate before turning it off. This is the pattern followed by alpha DRA feature gates in k8s as well.

Hence we guard users at webhook level and then leave it up to them to handle turning off the gate with care.

Reply:
synced with @jean-edouard on Slack, got an agreement that the thought process here is valid.

Reply:
My point remains that the feature gate should be checked on VM start, not creation.
I agree that we don't want to fully support dynamic feature gates switches and all the edge cases that come with it, but I still disagree with having the check in a webhook, i.e. on creation. It should be just as simple, and much more elegant, to reject VM start when the necessary FG is not set.

Reply:
VM start and VMIs creation?

Reply:
Good point yeah, and since a VM start is effectively a VMI creation, we could indeed make it all work just by adding to the VMI admission webhook.
However, we try not to add webhooks that depend on external components, which is why I think the VMI controller is a better location for the check.

Reply:
+1 VM start and VMIs creation works for me, I will add that to the document